File: backend/app/schemas/tstv.py
"""Pydantic schemas for TSTV and DRM endpoints."""

import uuid
from datetime import datetime

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# DRM
# ---------------------------------------------------------------------------

class DRMKeyResponse(BaseModel):
    key_id: uuid.UUID
    key_value_hex: str
    channel_key: str
    active: bool


class ClearKeyLicenseRequest(BaseModel):
    kids: list[str]
    type: str = "temporary"


class ClearKeyLicenseResponse(BaseModel):
    keys: list[dict]


# ---------------------------------------------------------------------------
# TSTV Channels
# ---------------------------------------------------------------------------

class TSTVChannelResponse(BaseModel):
    id: uuid.UUID
    name: str
    cdn_channel_key: str
    tstv_enabled: bool
    startover_enabled: bool
    catchup_enabled: bool
    cutv_window_hours: int

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Start Over
# ---------------------------------------------------------------------------

class ScheduleEntrySummary(BaseModel):
    id: uuid.UUID
    channel_id: uuid.UUID
    title: str
    synopsis: str | None = None
    genre: str | None = None
    start_time: datetime
    end_time: datetime
    catchup_eligible: bool
    startover_eligible: bool

    model_config = {"from_attributes": True}


class StartOverAvailability(BaseModel):
    channel_id: uuid.UUID
    schedule_entry: ScheduleEntrySummary
    startover_available: bool
    elapsed_seconds: int


# ---------------------------------------------------------------------------
# Catch-Up
# ---------------------------------------------------------------------------

class CatchUpProgram(BaseModel):
    schedule_entry: ScheduleEntrySummary
    expires_at: datetime
    bookmark_position_seconds: int | None = None


class CatchUpListResponse(BaseModel):
    programs: list[CatchUpProgram]
    total: int


class CatchUpByDateResponse(BaseModel):
    programs: list[CatchUpProgram]
    total: int
    date: str


# ---------------------------------------------------------------------------
# TSTV Sessions
# ---------------------------------------------------------------------------

class TSTVSessionCreate(BaseModel):
    channel_id: str
    schedule_entry_id: uuid.UUID
    session_type: str = Field(pattern=r"^(startover|catchup)$")
    profile_id: uuid.UUID | None = None


class TSTVSessionUpdate(BaseModel):
    last_position_s: float | None = Field(default=None, ge=0)
    completed: bool | None = None


class TSTVSessionResponse(BaseModel):
    id: int
    channel_id: str
    schedule_entry_id: uuid.UUID
    session_type: str
    started_at: datetime
    last_position_s: float
    completed: bool

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Admin — TSTV Rules
# ---------------------------------------------------------------------------

class TSTVRulesResponse(BaseModel):
    channel_id: uuid.UUID
    channel_name: str
    tstv_enabled: bool
    startover_enabled: bool
    catchup_enabled: bool
    cutv_window_hours: int
    catchup_window_hours: int


class TSTVRulesUpdate(BaseModel):
    tstv_enabled: bool | None = None
    startover_enabled: bool | None = None
    catchup_enabled: bool | None = None
    cutv_window_hours: int | None = Field(default=None)


# ---------------------------------------------------------------------------
# Admin — SimLive
# ---------------------------------------------------------------------------

class SimLiveChannelStatus(BaseModel):
    channel_key: str
    running: bool
    pid: int | None = None
    segment_count: int = 0
    disk_bytes: int = 0
    error: str | None = None


class CleanupResult(BaseModel):
    channels_processed: int
    total_segments_deleted: int
    total_bytes_freed: int

File: frontend-client/src/api/tstv.ts
import { apiFetch } from './client'

// ---------------------------------------------------------------------------
// TSTV Channels
// ---------------------------------------------------------------------------

export interface TSTVChannel {
  id: string
  name: string
  cdn_channel_key: string
  tstv_enabled: boolean
  startover_enabled: boolean
  catchup_enabled: boolean
  cutv_window_hours: number
}

export function fetchTSTVChannels(): Promise<TSTVChannel[]> {
  return apiFetch<TSTVChannel[]>('/tstv/channels')
}

// ---------------------------------------------------------------------------
// Start Over
// ---------------------------------------------------------------------------

export interface ScheduleEntrySummary {
  id: string
  channel_id: string
  title: string
  synopsis: string | null
  genre: string | null
  start_time: string
  end_time: string
  catchup_eligible: boolean
  startover_eligible: boolean
}

export interface StartOverAvailability {
  channel_id: string
  schedule_entry: ScheduleEntrySummary
  startover_available: boolean
  elapsed_seconds: number
}

export function getStartOverAvailability(channelId: string): Promise<StartOverAvailability> {
  return apiFetch<StartOverAvailability>(`/tstv/startover/${channelId}`)
}

export function getStartOverManifestUrl(channelId: string, scheduleEntryId: string): string {
  const base = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000/api/v1'
  return `${base}/tstv/startover/${channelId}/manifest?schedule_entry_id=${scheduleEntryId}`
}

// ---------------------------------------------------------------------------
// Catch-Up
// ---------------------------------------------------------------------------

export interface CatchUpProgram {
  schedule_entry: ScheduleEntrySummary
  expires_at: string
  bookmark_position_seconds: number | null
}

export interface CatchUpListResponse {
  programs: CatchUpProgram[]
  total: number
}

export interface CatchUpByDateResponse {
  programs: CatchUpProgram[]
  total: number
  date: string
}

export function listCatchUpPrograms(
  channelId: string,
  limit = 50,
  offset = 0,
): Promise<CatchUpListResponse> {
  const params = new URLSearchParams({ limit: String(limit), offset: String(offset) })
  return apiFetch<CatchUpListResponse>(`/tstv/catchup/${channelId}?${params}`)
}

export function listCatchUpByDate(
  date?: string,
  channelId?: string,
  genre?: string,
  limit = 50,
  offset = 0,
  profileId?: string,
): Promise<CatchUpByDateResponse> {
  const params = new URLSearchParams({ limit: String(limit), offset: String(offset) })
  if (date) params.set('date', date)
  if (channelId) params.set('channel_id', channelId)
  if (genre) params.set('genre', genre)
  if (profileId) params.set('profile_id', profileId)
  return apiFetch<CatchUpByDateResponse>(`/tstv/catchup?${params}`)
}

export function getCatchUpManifestUrl(channelId: string, scheduleEntryId: string): string {
  const base = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000/api/v1'
  return `${base}/tstv/catchup/${channelId}/manifest?schedule_entry_id=${scheduleEntryId}`
}

// ---------------------------------------------------------------------------
// DRM ClearKeys
// ---------------------------------------------------------------------------

export function getClearKeys(channelId: string): Promise<{ keys: Record<string, string> }> {
  return apiFetch<{ keys: Record<string, string> }>(`/drm/clearkeys/${channelId}`)
}

// ---------------------------------------------------------------------------
// TSTV Sessions
// ---------------------------------------------------------------------------

export interface TSTVSession {
  id: number
  channel_id: string
  schedule_entry_id: string
  session_type: 'startover' | 'catchup'
  started_at: string
  last_position_s: number
  completed: boolean
}

export function createTSTVSession(
  channelId: string,
  scheduleEntryId: string,
  sessionType: 'startover' | 'catchup',
  profileId?: string,
): Promise<TSTVSession> {
  return apiFetch<TSTVSession>('/tstv/sessions', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      channel_id: channelId,
      schedule_entry_id: scheduleEntryId,
      session_type: sessionType,
      profile_id: profileId ?? null,
    }),
  })
}

export function updateTSTVSession(
  sessionId: number,
  lastPositionS?: number,
  completed?: boolean,
): Promise<TSTVSession> {
  const body: Record<string, unknown> = {}
  if (lastPositionS !== undefined) body.last_position_s = lastPositionS
  if (completed !== undefined) body.completed = completed
  return apiFetch<TSTVSession>(`/tstv/sessions/${sessionId}`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  })
}
